libs/stokesian_dynamics/device_matrix.hpp

```diff
@@ -0,0 +1,788 @@
+    using vector = thrust_wrapper::device_vector<T>;
+    static std::remove_const<decltype(thrust_wrapper::device)>::type par()
+      { return thrust_wrapper::device; }
+};
+
+template <typename...>
+using void_t = void;
+
+template <typename Policy, typename = void>
+struct is_policy : std::false_type {};
+
+template <typename Policy>
+struct is_policy<Policy, void_t<typename Policy::template vector<class T>,
+                                decltype(Policy::par())>> : std::true_type {};
+
+} // namespace policy
+
+DEVICE_FUNC inline thrust_wrapper::tuple<std::size_t, std::size_t>
+unravel_index(std::size_t index, std::size_t lda) {
+    std::size_t i, j;
+    i = index % lda;
+    index /= lda;
+    j = index;
+    return thrust_wrapper::make_tuple(i, j);
+    ;
+}
+
+/// \cond
+
+// LAPACK prototypes
+#ifndef __CUDACC__
+extern "C" {
+int dgemm_(char *transa, char *transb, int *m, int *n, int *k, double *alpha,
+           double *a, int *lda, double *b, int *ldb, double *beta, double *c,
+           int *ldc);
+
+int dgemv_(char *trans, int *m, int *n, double *alpha, double *a, int *lda,
+           double *x, int *incx, double *beta, double *y, int *incy);
+
+int dpotrf_(char *uplo, int *n, double *a, int *lda, int *info);
+
+int dpotrs_(char *uplo, int *n, int *nrhs, double *a, int *lda, double *b,
+            int *ldb, int *info);
+}
+#endif
+
+/** The `cublas` and `cusolver` structs channel access to efficient matrix
+ *  operations provided by either the cuBLAS library (after which it is named),
+ *  which executes on the GPU, or by BLAS/LAPACK, which execute on the CPU.
+ *  The first template parameter specifiies whether the routines are executed on
+ *  host or on device, by either passing `policy::host` or `policy::device`.
+ *  The second template parameter specifies the data type (only `double` is 
+ *  available).
+ */
+/** The `cublas` struct channels access to efficient basic matrix operations
+ *  provided by either the cuBLAS library (after which it is named), which
+ *  executes on the GPU, or by BLAS (Basic Linear Algebra Subprograms), which
+ *  executes on the CPU.
+ *  The first template parameter specifiies whether the routines are executed on
+ *  host or on device, by either passing `policy::host` or `policy::device`.
+ *  The second template parameter specifies the data type (only `double` is 
+ *  available).
+ */
+namespace internal {
+
+/** The `cublas` struct channels access to efficient basic matrix operations
+ *  provided by either the cuBLAS library (after which it is named), which
+ *  executes on the GPU, or by BLAS (Basic Linear Algebra Subprograms), which
+ *  executes on the CPU.
+ *  The first template parameter specifies whether the routines are executed on
+ *  host or on device, by either passing `policy::host` or `policy::device`.
+ *  The second template parameter specifies the data type (only `double` is 
+ *  available).
+ */
+template <typename, typename>
+struct cublas {};
```

Comment:
maybe this templated struct documentation could be improved with `\tparam`, but I don't have a strong opinion on that one given that Doxygen will not parse it (it's inside a `\cond` block)
```suggestion
 *  \tparam Policy Specifies whether the routines are executed on host or on
 *                 device, by either passing \ref policy::host or \ref
 *                 policy::device
 *  \tparam T      Specifies the data type (only `double` is available).
 */
template <typename Policy, typename T>
struct cublas {};
```